clang/lib/Sema/BoundsAnalysis.cpp

```diff
@@ -0,0 +1,491 @@
+//===--------- BoundsAnalysis.cpp - Bounds Widening Analysis --------------===//
+//
+//                     The LLVM Compiler Infrastructure
+//
+// This file is distributed under the University of Illinois Open Source
+// License. See LICENSE.TXT for details.
+//
+//===----------------------------------------------------------------------===//
+//
+//  This file implements a dataflow analysis for bounds widening.
+//
+//===----------------------------------------------------------------------===//
+
+#include "clang/Sema/BoundsAnalysis.h"
+
+namespace clang {
+
+void BoundsAnalysis::WidenBounds() {
+  assert(Cfg && "expected CFG to exist");
+
+  WorkListTy WorkList;
+  BlockMapTy BlockMap;
+
+  // Add each block to WorkList and create a mapping from Block to
+  // ElevatedCFGBlock.
+  // Note: By default, PostOrderCFGView iterates in reverse order. So we always
+  // get a reverse post order when we iterate PostOrderCFGView.
+  for (const auto *B : PostOrderCFGView(Cfg)) {
+    // SkipBlock will skip all null, entry and exit blocks. PostOrderCFGView
+    // does not contain any unreachable blocks. So at the end of this loop
+    // BlockMap only contains reachable blocks.
+    if (SkipBlock(B))
+      continue;
+
+    auto EB = new ElevatedCFGBlock(B);
+    // Note: WorkList is a queue. So we maintain the reverse post order when we
+    // iterate WorkList.
+    WorkList.append(EB);
+    BlockMap[B] = EB;
+  }
+
+  // At this time, BlockMap only contains reachable blocks. We iterate through
+  // all blocks in the CFG and append all unreachable blocks to the WorkList.
+  for (auto I = Cfg->begin(), E = Cfg->end(); I != E; ++I) {
+    const auto *B = *I;
+    if (!SkipBlock(B) && !BlockMap.count(B)) {
+      auto EB = new ElevatedCFGBlock(B);
+      WorkList.append(EB);
+      BlockMap[B] = EB;
+    }
+  }
+
+  // Compute Gen and Kill sets.
+  ComputeGenSets(BlockMap);
+  ComputeKillSets(BlockMap);
+
+  // Compute In and Out sets.
+  while (!WorkList.empty()) {
+    auto *EB = WorkList.next();
+    WorkList.remove(EB);
+
+    ComputeInSets(EB, BlockMap);
+    ComputeOutSets(EB, BlockMap, WorkList);
+  }
+
+  CollectWidenedBounds(BlockMap);
+}
+
+void BoundsAnalysis::ComputeGenSets(BlockMapTy BlockMap) {
+  // If there is an edge B1->B2 and the edge condition is of the form
+  // "if (*(p + i))" then Gen[B1] = {B2, p:i} .
```

Comment:
Could you clarify whether i is a constant or a variable?   I know that i is a constant from context, but it could confuse people later on.

Gen sets should be constant offsets relative to the declared upper bound of a pointer.     If p has declared upper bounds of e, and this is a dereference of *(e + i), then Gen[B1] = i.   You'll need to handle expressions that are equivalent to `e + i`.  Note that we already have methods for checking the equivalence of expressions.